### PR TITLE
Add "GSoC 2024: The 1001 thresholds in LLVM" blog post

### PR DESCRIPTION
**Pull request by Sh0g0-1758**
A blog post briefly outlining the work I had done over the course of GSoC 2024.
cc: @asl

### CODE REVIEW

**Review by asl — CHANGES_REQUESTED** [review]
The objectives & deliverables for this project as states in the proposal are:

Objectives:
-   Explore different thresholds in llvm.
-   Understanding what it means for a threshold to be hit.
-   Profiling different thresholds.
-   Selecting optimal values for different thresholds.

Deliverables:
-    Statistical evidence on the impact of various thresholds inside of LLVM’s code base, including compile time changes, impact on transformations, and performance measurements.

Ensure that you expand properly both objectives and deliverables and present their properly. E.g. I do not see "statistical evidence", provide detailed information about "impact of various thresholds inside of LLVM’s code base, including compile time changes, impact on transformations, and performance measurements"

It seems you produces lots of graphs, but no non-trivial results were provides. Ensure you addressed the feedback you had on Discourse.

**Comment by Sh0g0-1758** [comment]
@asl, would you like me to add more or are the changes enough ?

File: content/posts/2024-08-31-the-1001-thresholds-in-llvm.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+We also experimented with varying thresholds cumulatively by leveraging file-specific information through an LLVM pass. However, after discussions with the mentors, this approach was set aside due to the significant changes it would necessitate across other parts of the LLVM codebase.
+
+As a result, we have not yet categorized different thresholds, such as identifying optimal threshold values for specific file types (e.g., I/O-intensive files). Nonetheless, we provide substantial [statistical evidence](https://sh0g0-1758.github.io/GSOC-X-LLVM/) in the form of LLVM statistics, bitcode size, and compile-time graphs, as well as histograms that examine these variations on a per-file basis. Additionally, a correlation table between knob values and performance metrics further illustrates the significant impact this study could have on improving LLVM's overall performance.
```
Comment (asl) [inline content/posts/2024-08-31-the-1001-thresholds-in-llvm.md:52]: I do not see any statistical evidence on the link provided in the common sense of this word. Just bunch of numbers without any conclusion, reasoning and analysis. How these results could be used? In some sense they are trivial and more or less well-known to the community.

**Comment by asl** [comment]
Tagging @jdoerfert @wsmoses @jhueckelheim

File: content/posts/2024-08-31-the-1001-thresholds-in-llvm.md (hunk 1 above)
Comment (Sh0g0-1758) [reply content/posts/2024-08-31-the-1001-thresholds-in-llvm.md]: @asl , I understand what you are saying but we never got to that part, it was mostly research to develop some experimental setup to collect these results.

cc: @jdoerfert @jhueckelheim

**Comment by jdoerfert** [comment]
> The objectives & deliverables for this project as states in the proposal are:
> 
> Objectives:
> 
> * Explore different thresholds in llvm.

There are spreadsheets listing lots of them and the ways they have been identified is documented.

> * Understanding what it means for a threshold to be hit.

We did this only in passing. Priorities shifted as the initial exploration of thresholds did yield way more than we expected. We have categorized them though, depending on how useful the next steps will be for them.

> * Profiling different thresholds.

That happened. And more importantly we have scripts to do that automatically and capture the data.

> * Selecting optimal values for different thresholds.

This one is tricky. We can read "optimal" from the graphs we have just fine, and we suggested as much on discourse. Thus, we can technically check this box. However, as we actually got the data in we pivoted towards more complex analysis to determine optima for different objective functions. That was the last step, work in progress, we did.


> Deliverables:
> 
> * Statistical evidence on the impact of various thresholds inside of LLVM’s code base, including compile time changes, impact on transformations, and performance measurements.

The graphs clearly show that. 10k IR files and for some of the knobs clear patterns emerge. Plus the scripts and data behind the graphs is available for alternative processing.

> Ensure that you expand properly both objectives and deliverables and present their properly. E.g. I do not see "statistical evidence", provide detailed information about "impact of various thresholds inside of LLVM’s code base, including compile time changes, impact on transformations, and performance measurements"
> 
> It seems you produces lots of graphs, but no non-trivial results were provides. Ensure you addressed the feedback you had on Discourse.

I'm not sure I'd agree that all of these are "trivial results", for most interpretations of that term:
- Identifying the knobs across the entire code base, not only a single "source" (like command line options),
- categorizing them according to usefulness in tuning,
- scripts that scale out to collect thousands of data points for 10k IR files,
- statistical correlation analysis (e.g., the correlation analysis between knob value and LLVM reported statistics found at the bottom of the pages: https://sh0g0-1758.github.io/GSOC-X-LLVM/graph10000.html?file=jump-threading-threshold.json),
- visualization in interactive graphs and tables of the collected data,
- best-case analysis (oracle-recommended knob values) for various knobs to show that there is not a single best value

File: content/posts/2024-08-31-the-1001-thresholds-in-llvm.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
+---
+author: "Shourya Goel"
+date: "2024-08-31"
```
Comment (asl) [inline content/posts/2024-08-31-the-1001-thresholds-in-llvm.md:3]:
```suggestion
date: "2024-10-21"
```

**Pull request merged**